libs/parallelism/execution/include/hpx/execution/sender_future.hpp

```diff
@@ -0,0 +1,137 @@
+#include <hpx/allocator_support/internal_allocator.hpp>
+#include <hpx/execution/algorithms/detail/single_result.hpp>
+#include <hpx/execution_base/operation_state.hpp>
+#include <hpx/execution_base/sender.hpp>
+#include <hpx/functional/invoke_result.hpp>
+#include <hpx/futures/detail/future_data.hpp>
+#include <hpx/futures/promise.hpp>
+#include <hpx/memory/intrusive_ptr.hpp>
+#include <hpx/type_support/pack.hpp>
+#include <hpx/type_support/unused.hpp>
+
+namespace hpx { namespace execution { namespace experimental {
+    namespace detail {
+        template <typename T, typename Allocator, typename OS>
+        struct future_data
+          : hpx::lcos::detail::future_data_allocator<T, Allocator>
+        {
+            HPX_NON_COPYABLE(future_data);
+
+            using operation_state_type = std::decay_t<OS>;
+            using init_no_addref =
+                typename hpx::lcos::detail::future_data_allocator<T,
+                    Allocator>::init_no_addref;
+            using other_allocator = typename std::allocator_traits<
+                Allocator>::template rebind_alloc<future_data>;
+
+            operation_state_type os;
+
+            template <typename S, typename R>
+            future_data(init_no_addref no_addref, other_allocator const& alloc,
+                S&& s, R&& r)
+              : hpx::lcos::detail::future_data_allocator<T, Allocator>(
+                    no_addref, alloc)
+              , os(hpx::execution::experimental::connect(
+                    std::forward<S>(s), std::forward<R>(r)))
+            {
+                hpx::lcos::detail::intrusive_ptr_add_ref(this);
+                hpx::execution::experimental::start(os);
+            }
+        };
+
+        template <typename T, typename Allocator>
+        struct make_future_receiver
+        {
+            hpx::lcos::detail::future_data_allocator<T, Allocator>* data;
+
+            void set_error(std::exception_ptr e) noexcept
+            {
+                data->set_exception(e);
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+
+            void set_done() noexcept
+            {
+                std::terminate();
+            }
+
+            template <typename U>
+            void set_value(U&& u) noexcept
+            {
+                data->set_value(std::forward<U>(u));
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+        };
+
+        template <typename Allocator>
+        struct make_future_receiver<void, Allocator>
+        {
+            hpx::lcos::detail::future_data_allocator<void, Allocator>* data;
+
+            void set_error(std::exception_ptr e) noexcept
+            {
+                data.set_exception(e);
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+
+            void set_done() noexcept
+            {
+                std::terminate();
+            }
+
+            template <typename Ts>
+            void set_value() noexcept
```

Comment:
Does this have to be a template, still?

Reply:
Nope (it can't be one), good catch! The void case is still untested, so I'll update this along with adding those tests.